Add SPACEXLAUNCH as a supported price identifier

Alirun opened the pull request.

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+| UMIP-XX                |                                                               |
+| ------------------- | ------------------------------------------------------------- |
+| UMIP Title          | Add SPACEXLAUNCH as a supported price identifier |
```
Very very interesting idea

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+- Price identifier name: **SPACEXLAUNCH** 
+- Markets & Pairs: **N/A**
+- Example data providers: https://www.rocketlaunch.live/?filter=spacex , https://en.wikipedia.org/wiki/List_of_Falcon_9_and_Falcon_Heavy_launches
```
How do we verify the authority of the source data? These websites could miss data, or report false data.

pemulis replied:
It would be great to have a few more data sources listed here, if possible, and industry sources to fact check Wikipedia and the like in the event of a dispute.

rollodex replied:
I think that multiple sources are definitely needed for repudiation.
Here is one that I found: https://docs.spacexdata.com/

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- Price identifier name: **SPACEXLAUNCH**
+- Base Currency: **N/A**
+- Quote Currency: **N/A**
+- Rounding: *Round to 2 decimal places (third decimal place digit >= 5 rounds up and < 5 rounds down)*
```
What would this rounding do? Is this the percentage of launch success rate?

smb2796 replied:
Rounding is still part of the template, so it seems fine to me.

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- Launch id - Name of the SpaceX mission
+- Weight (`Wi`) - Weight of the launch in the overall success rate calculation
+
+Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions
```
What about launches that didn't intend for the rocket to land? I feel like this ancillary data requires a review from the rocket domain expert. Where did you derive this success rate formula `Success Rate =  SUM(Si * Wi) / SUM(Wi)`?

smb2796 replied:
I think it is a good point, but I see this as being more of a product question for the contract deployer around what launches are included in the ancillary data list. I don't think this has any impact on our ability to return a price as specified though.

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions
+- If launch did not happen till date or the rocket was not successfully lifted off, then `Si=0`
+- If the rocket successfully lifted off, but never landed, or the landing was not successful then `Si=0.5`
```
Again, questioning why 0.5? The launch is difficult, and some may not intend to land.

mhairimca replied:
If it lands on earth, or anywhere?  What if it docks at a space station, is that considered "landing"?

smb2796 replied:
Agreed that the criteria for each bucket could use some additional clarification

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- If the rocket successfully lifted off, but never landed, or the landing was not successful then `Si=0.5`
+- If the rocket successfully lifted off, and landed, then `Si=1`
+
+Liftoff is considered successful if the rocket reached expected altitude without any significant damage and performed the required part of the mission at this altitude.
```
We need to further define `expected altitude` as the rockets do see explosions even after reaching significant altitude. All these definitions should link a reference the authoritative source.

ChandlerDeKock replied:
Agree with this comment. The rocket will reach its expected altitude and then be deemed a success but then blow up later, which for an insurance idea might not be useful. I am also not sure what significant damage means?

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+Liftoff is considered successful if the rocket reached expected altitude without any significant damage and performed the required part of the mission at this altitude.
+
+Landing is considered successful if the rocket landed on Earth without any significant damage till date.
```
Why limited on `Earth`?

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+- Request timestamp: 1614556800 (March 1st 2021 00:00 UTC)
+- List of launches:
+- - Launch 1: ID: "Starlink-18”, W: 1
```
You mean `Wi`?

pemulis replied:
I think `i` is the index (so `1` here and `2` for the next one).

therealhexi replied:
Make sense

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+---
+
+## Description
+Ancillary data would consist of the list of launches, where each launch is described by
```
How do we determine the list? Do we always use the beginning of time as the start filter date, and contract expiration as the end filter date?

smb2796 replied:
Good questions! The list is passed in an ancillary data and Ali also updated to clarify that the end timestamp is the price request timestamp

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+### Example 2
+- Request timestamp: 1612137600 (February 1st 2021 00:00 UTC)
+- List of launches:
+- -  Launch 1: ID: "Transporter-1”, W: 1
```
`Wi`

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
````diff
@@ -0,0 +1,118 @@
+
+```
+[{
+  id: string,
````
I can foresee future where we want to update different rocket type with different parameters. Not all launches are equal.

pemulis replied:
I think that's what the `weight` field is for, so you can mark certain launches as more or less important to your calculation.

therealhexi replied:
My understanding was `weight` was used for the actual `weight of the rocket` lol. Fair, we should have a clear definition of the weight in this case

therealhexi commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+# Rationale
+
+No rationale is needed. The motivation for this price identifier is explained in motivation.
```
I would argue rationale for the definition, and formula selection should be included here. As a first time reader, I'm not convinced this is the right implementation (although I agree this is a very very innovative idea). Let's cite some sources where we get these formula

therealhexi reviewed (commented):
Initial review

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+# Summary 
+
+The DVM should support price requests for SPACEXLAUNCH. The purpose of the SPACEXLAUNCH price identifier is to be a generalized way to evaluate the success rate of the rocket launches made by SpaceX till date considering provided ancillary data
```
Until a specified date? Is the number a cumulative score increasing over time with the number of SpaceX launches? Would the price identifier consider all of the launches prior to the creation of the identifier? If so, it might be useful to show the current score along with a table of SpaceX's launch history matched against their score and weight, as an example of how the identifier works.

chrismaree replied:
```suggestion
The DVM should support price requests for SPACEXLAUNCH. The purpose of the SPACEXLAUNCH price identifier is to be a generalized way to evaluate the success rate of the rocket launches made by SpaceX till date considering provided ancillary data.
```

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+# Motivation
+
+Provision of the way to evaluate the success rate of SpaceX launches would open a wide range of possible synthetics based on it (e.g. insurances)
```
`Adding a way to evaluate the success rate of SpaceX launches would open a wide range of potential synthetics (e.g. insurance).`

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+# Price Feed Implementation
+
+Price Feed Implementation is unnecessary for this price identifier, as it will not have liquidation bots that need to programmatically get this price.
```
Will this be calculated by hand on the expiry of a contract?

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
````diff
@@ -0,0 +1,118 @@
+
+Success rate than can be calculated by
+```
+Success Rate =  SUM(Si * Wi) / SUM(Wi)
````
Just to check my understanding, `Wi` would be passed in through ancillary data, and what the price proposer will need to provide is the result of the calculation after adding `Si` for each of the relevant launches?

smb2796 replied:
This matches my understanding

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+Launch 1 happened till date, successfully lifted off and landed: `S1=1`
+
+Launch 2 happened till date, successfully lifted off, but did not landed: `S1=0.5`
```
`S2=0.5`, I think.

pemulis replied:
`Launch 2 happened, successfully lifted off, but did not land`

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- - Launch 1: ID: "Starlink-18”, W: 1
+- - Launch 2: ID: "Starlink-19”, W: 1
+
+Launch 1 happened till date, successfully lifted off and landed: `S1=1`
```
`Launch 1 happened, successfully lifted off, and landed`

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- List of launches:
+- -  Launch 1: ID: "Transporter-1”, W: 1
+
+Launch 1 happened till date, successfully lifted off and landed: S1=1
```
`Launch 1 happened, successfully lifted off, and landed`

pemulis commented on UMIPs/spacex_launch_price_identifier.md:
````diff
@@ -0,0 +1,118 @@
+```
+
+##  Format
+Ancillary data in the request will be passed as a JSON  in the following format
````
Extra space here.

Reinis-FRP commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+i.e. sum of all products of launch statuses and corresponding weights divided by the sum of all weights
+
+## Examples
```
I believe Examples belong more to the Implementation section

Reinis-FRP commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- Launch id - Name of the SpaceX mission
+- Weight (`Wi`) - Weight of the launch in the overall success rate calculation
+
+Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions
```
The formula and conditions for `Si` calculation seems more belonging to the Implementation section

smb2796 replied:
Bump. @Alirun can you move the description/examples to the implementation section?

mhairimca commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+## Description
+Ancillary data would consist of the list of launches, where each launch is described by
+- Launch id - Name of the SpaceX mission
+- Weight (`Wi`) - Weight of the launch in the overall success rate calculation
```
Whats the difference between Wi and Si?

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+Ancillary data in the request will specify the parameters that should be considered during the oracle result determination process.
+
+# Motivation
```
I think it is worth noting that this price identifier is not intended to be used with a registered UMA financial contract type. Is it possible to link to the Opium contract that will use this price id?

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+
+# Rationale
+
+No rationale is needed. The motivation for this price identifier is explained in motivation.
```
Agreed. The results of this price are manipulatable without some criteria around the weighting associated with each launch.

smb2796 replied:
bump: @Alirun can you add some clarification around what how "weights" are chosen? This isn't incredibly necessary, but would help voters determine when ancillary data requests were submitted incorrectly.

Alirun commented:
Addressed PR comments:
- Clarifications on timing and request date
- Additional motivation text
- Added data consumer example
- Additional datasource
- Clarification on ancillary data purpose and structure
- Fixed typos and errors
- Changed JSON format to more compact one
- Added ancillary data parsing example

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,151 @@
+- Weight (`Wi`) - Mathematical `weight` of the launch in the overall success rate calculation. Specified wight might be any positive number that is used to define the importance of this particular launch in the overall success rate respectively to other launches provided in ancillary data.
+
+Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions
+- If launch did not happen until request date or the rocket was not successfully lifted off, then `Si=0`
```
```suggestion
- If launch did not happen until request date or the rocket did not successfully lift off, then `Si=0`
```

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+Each launch in the list should be evaluated and its status (`Si`) should be determined according to these conditions
+- If launch did not happen till date or the rocket was not successfully lifted off, then `Si=0`
+- If the rocket successfully lifted off, but never landed, or the landing was not successful then `Si=0.5`
+- If the rocket successfully lifted off, and landed, then `Si=1`
```
@Alirun can you also add more explicit language around this criteria in relation to the price request timestamp, like @Reinis-FRP said? Simply a "landing was not successful before the price request timestamp", and "liftoff was not successful before the price request timestamp"

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,118 @@
+- Markets & Pairs: **N/A**
+- Example data providers: https://www.rocketlaunch.live/?filter=spacex , https://en.wikipedia.org/wiki/List_of_Falcon_9_and_Falcon_Heavy_launches
+- Cost to use: **None**
+- Real-time data update frequency: **Daily**
+- Historical data update frequency: **Daily**
```
Bump: @Alirun can you update this to `NA`?

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,151 @@
+## Description
+
+Ancillary data are provided by price requesters and would consist of the list of launches, where each launch is described by
+- Launch id - Name of the SpaceX mission, that allows to uniquely identify particular launch.
```
```suggestion
- Launch id - Name of the SpaceX mission, that allows voters to uniquely identify a particular launch.
```

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
````diff
@@ -0,0 +1,151 @@
+```
+
+## Format
+Ancillary data in the request will be passed as a JSON  in the following format
````
@Alirun can you also add some clarification here for a fallback in situations where ancillary data cannot be identified?

Something like:

```
If voters cannot correctly match a specific ID to an individual launch, `Si` should equal 0 for that launch.

If voters cannot correctly decode the ancillary data into the format defined below, voters should return 0.5.
```

Doesn't necessarily need to be 0.5, but we need some contingency instructions for situations where ancillary data is mis-parameterized.

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,151 @@
+- - Launch 1: `{ "id": "Starlink-18", "w": 1 }`
+- - Launch 2: `{ "id": "Starlink-19", "w": 1 }`
+
+Launch 1 happened until request date, successfully lifted off, and landed: `S1=1`
```
```suggestion
Launch 1 happened before request date, successfully lifted off, and landed: `S1=1`
```

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,151 @@
+
+Launch 1 happened until request date, successfully lifted off, and landed: `S1=1`
+
+Launch 2 happened until request date, successfully lifted off, but did not land: `S2=0.5`
```
```suggestion
Launch 2 happened before request date, successfully lifted off, but did not land: `S2=0.5`
```

smb2796 commented on UMIPs/spacex_launch_price_identifier.md:
```diff
@@ -0,0 +1,151 @@
+- List of launches:
+- -  Launch 1: `{ "id": "Transporter-1", "w": 1 }`
+
+Launch 1 happened until request date, successfully lifted off, and landed: `S1=1`
```
```suggestion
Launch 1 happened before request date, successfully lifted off, and landed: `S1=1`
```

smb2796 commented:
Looks good overall, thanks for making those updates!

Outstanding issues that still should be addressed, but I can fix on my side since they are mostly just clarification/formatting issues.

1. Moving implementation information to the information section.
2. Changing update frequency away from `Daily`
3. Clarification around the bounds of the price request timestamp being equivalent to "never landing"
4. Explanation into Rationale around why weights are being chosen - as explained on the UMIP call, this is because it allows product creators to create differently balanced insurance products.

Going to approve and merge into Last Call and address these smaller issues. @Alirun will tag you on the pr to make these.
